rfcs/001-alternate-crypto-backends.md

```diff
@@ -0,0 +1,100 @@
+| Key exchange | Statically uses *ring* | X25519 | `src/kx.rs` |
+| TLS1.2 PRF | Statically uses *ring* | HMAC-SHA256 | `src/tls12/kdf.rs` |
+| TLS1.3 key schedule | Statically uses *ring* | HKDF-HMAC-SHA256 | `src/tls13/key_schedule.rs` |
+| Server ticket production | Pluggable extension point. Default does not support ticket generation. Provided optional implementation uses *ring*. | CHACHA20-POLY1305 | `src/ticketer.rs` |
+
+For the items here that statically depend on *ring*, how do we determine what
+individual algorithm to use?
+
+- Random material generation: currently fixed.
+- Bulk encryption/decryption: from cipher suite (`SupportedCipherSuite`)
+- Key exchange: from key exchange group (`SupportedKxGroup`)
+- TLS1.2 PRF: from cipher suite (`SupportedCipherSuite::Tls12`)
+- TLS1.3 key schedule: from cipher suite (`SupportedCipherSuite::Tls13`)
+
+The principal observation here is: the same mechanism which means that AES does not appear
+present in an application binary which only mentioned `TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256`
+should be the same one that lets an application avoid a *ring* dependency.
+
+This gives us five remaining issues to deal with:
+
+### Random generation
+
+This is needed early on, so we cannot hook it to anything else (like the cipher suite, say, but
+that would be odd anyway.)
+
+We should make this configurable in the `ServerConfig`/`ClientConfig` types.
+
+### Bulk encryption/decryption
+
+We have suitable traits for this already, rooted in `SupportedCipherSuite::Tls12`.
+The same would need doing for `SupportedCipherSuite::Tls13`.
+
+### Key exchange
+`SupportedKxGroup` should contain a functor which returns a trait implementation that
+replaces the *ring*-oriented functions in the `KeyExchange` struct.
+
+### TLS1.2 PRF and TLS1.3 key schedule
+We should define a trait for incremental HMAC computation.
+
+In `SupportedCipherSuite` we should have a function that returns an implementation of that trait.
+
+This can be used by the TLS1.2 PRF and TLS1.3 key schedule.  The latter should wrap HMAC to maintain
+type safety of the HKDF uses.
+
+# Proposal
+
+## Crate organisation
+
+1. Move most of the crate to a `rustls-core` crate.  This should not depend on webpki, *ring* or sct.
+   But, alone, it is not useful: it does not contain any `SupportedCipherSuite`s, `SupportedKxGroup`s, etc.
+   However, it does contain all the traits and types for implementing these.
+2. `rustls` reexports most of the public parts of `rustls-core`, plus provides the default
+   `SupportedCipherSuite`s, `SupportedKxGroup`s, webpki-backed certificate verifier, loading and using
+   signing keys, early steps of config builders (as these know the default cipher suites), etc.
+
+   Most importantly: everything that `rustls` does can equally be done by a downstream crate against
+   different dependencies.
```

Comment:
It's really not obvious to me what we win by separating a rustls-core crate out from the rustls crate. If you're worried about coherence or visibility traps that make it impossible to implement the trait outside of rustls, I think we need a secondary implementation anyway to really flush those out but I don't think it is a strong argument for moving the trait implementation out of the rustls crate at this stage.

Reply:
My thought here was that _something_ needs to take the dependency on *ring*. If we have that in one crate, then we must rely on cargo features to enable/disable that dependency and that comes with the usual issue that (for a large enough application with enough dependencies) it's actually really annoying to disable a default feature. It seems easier to me for big applications to just say `[patch] rustls = { package = "rustls-wasi-crypto" }` or whatever.

Reply:
Hmm, it's an interesting idea. So far I've always considered `patch` entries to be temporary, so (as the maintainer of a semi-large application in my day job) I would hesitate to use that as a long term strategy. I also don't know how well it deals with version drift over time, can you `patch` rustls 0.22 and 0.23 separately?

Anyway, I would still maintain that we should work on the trait(s) first and consider splitting up the crate as more of an endgame thing, it's mostly orthogonal to the design anyway.

Alternatives: make `ring` not a default feature, so downstream libraries have to make an explicit choice to enable it, to raise the salience of crypto provider variance. Or maybe more far out, split out the `Config` types and builder into a separate crate which can maintain a more stable API (that is, 1.0), making it more feasible for downstream crates to expose that as part of their own public API.

Reply:
`patch` is a bad idea for production.

> make `ring` not a default feature, so downstream libraries have to make an explicit choice to enable it, to raise the salience of crypto provider variance.

What about requiring a `Crypto` trait to be passed to every builder?

Reply:
One alternative would be to standardize over the RustCrypto/traits and use https://crates.io/crates/ring-compat to fill in the ring version.